From 01e6453f37ae37340dc722032a2b1212c211698e Mon Sep 17 00:00:00 2001
From: Pavel Nedev <pnedev@mm-sol.com>
Date: Fri, 28 Jan 2011 14:17:19 +0200
Subject: OMAP4/S3D V4L2: Display change modifications

Disable previous display manager when changing displays.
Disable panel to be attached if being enabled.
Unbind overlay from manager only in case it was previously bound.
Mgr->apply removes as ovl->unset_manager will only happen if overlay was already disabled.
Checking if streaming is on is removed from change_display and done in the s_ctrl function itself.
Check if desired panel is the same as currently active is added to s_ctrl V4L2 function.

Change-Id: I3dad1e4c05e5b79d075692e169c846f47ae26db1
Signed-off-by: Pavel Nedev <pnedev@mm-sol.com>
---
 drivers/media/video/omap/omap_s3d_overlay.c |   23 +++++++++++++----------
 1 files changed, 13 insertions(+), 10 deletions(-)

diff --git a/drivers/media/video/omap/omap_s3d_overlay.c b/drivers/media/video/omap/omap_s3d_overlay.c
index 8fb0089..4d97d06 100644
--- a/drivers/media/video/omap/omap_s3d_overlay.c
+++ b/drivers/media/video/omap/omap_s3d_overlay.c
@@ -668,10 +668,8 @@ static int change_display(struct s3d_ovl_device *dev,
 		return -EINVAL;
 	}
 
-	if (dev->streaming) {
-		S3DERR("cannot change display while streaming!\n");
-		return -EBUSY;
-	}
+	if (dev->cur_disp == display)
+		return 0;
 
 	mgr = display->manager;
 	if (!mgr) {
@@ -692,7 +690,9 @@ static int change_display(struct s3d_ovl_device *dev,
 	}
 
 	if (mgr->device != display) {
+		omapdss_display_disable(display);
 		if (mgr->device) {
+			omapdss_display_disable(mgr->device);
 			r = mgr->unset_device(mgr);
 			if (r || mgr->device) {
 				S3DERR("failed to unbind display from manager\n");
@@ -744,13 +744,14 @@ static inline int bind_ovl2display(struct s3d_ovl_device *dev,
 		return r;
 	}
 
-	r = ovl->dssovl->unset_manager(ovl->dssovl);
-	if (r && mgr != NULL) {
-		S3DERR("couldn't unbind manager from ovl\n");
-		return r;
+	if (mgr) {
+		r = ovl->dssovl->unset_manager(ovl->dssovl);
+		if (r) {
+			S3DERR("couldn't unbind manager from ovl\n");
+			return r;
+		}
 	}
-	if (mgr)
-		mgr->apply(mgr);
+
 	r = ovl->dssovl->set_manager(ovl->dssovl, display->manager);
 	if (r) {
 		S3DERR("couldn't bind to manager\n");
@@ -2769,6 +2770,8 @@ static int vidioc_s_ctrl(struct file *file, void *fh, struct v4l2_control *a)
 			int r;
 			if (dev->streaming)
 				return -EBUSY;
+			if (dev->cur_disp_idx == a->value)
+				return 0;
 			mutex_lock(&dev->lock);
 			r = change_display(dev, get_display(a->value));
 			if (!r)
-- 
1.6.0.4

